include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +956,190 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
 #endif //NDF_PHONG
-	{}
+		CASE_OTHERWISE
+		{}
+		END_CASES
+
+		vec3 localL;
+		float VdotH = dot(localH, localV);
+		float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+		vec3 fr;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+		else
+#endif
+			fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
+		const float refractionProb = dot(CIE_XYZ_Luma_Y_coeffs, fr);
+		float rcpChoiceProb;
+		const bool refraction = is_bsdf ? irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb) : false;
+		rcpChoiceProb = is_bsdf ? rcpChoiceProb : 1.0;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			rem *= fr;
+#endif
+		pdf /= rcpChoiceProb;
+		float eta = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+		float rcpEta = 1.0/eta;
+
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(refraction, localV, localH, localL, rcpEta, rcpEta*rcpEta);
+		s = irr_glsl_createLightSampleTangentSpace(localV, localL, tangentFrame);
+
+		const float TdotH2 = out_microfacet.TdotH * out_microfacet.TdotH;
+		const float BdotH2 = out_microfacet.BdotH * out_microfacet.BdotH;
+		const float NdotH2 = out_microfacet.isotropic.NdotH2;
+		const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+		CASE_BEGIN(ndf, NDF_GGX)
+		{
+			G2_over_G1 = irr_glsl_ggx_smith_G2_over_G1(NdotL, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_GGXSmith_G1_wo_numerator(NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+			ndf_val = irr_glsl_ggx_aniso(TdotH2, BdotH2, NdotH2, ax, ay, ax2, ay2);
+		} CASE_END
+#endif //NDF_GGX
+
+#ifdef NDF_BECKMANN
+		CASE_BEGIN(ndf, NDF_BECKMANN)
+		{
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, ax2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
+#endif //NDF_BECKMANN
+
+#ifdef NDF_PHONG
+		CASE_BEGIN(ndf, NDF_PHONG)
+		{
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, s.NdotL * s.NdotL, ax2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
+#endif //NDF_PHONG
+		CASE_OTHERWISE
+		{}
+		END_CASES
+
+		rem *= G2_over_G1;
+		pdf *= irr_glsl_smith_VNDF_pdf_wo_clamps(ndf_val, G1_over_2NdotV);
+	} else
+#endif
+	{} //empty braces for `else`
+
+	out_remainder = rem; // TODO not multiply rem by weight??
+	out_pdf = weight*pdf; 
```

Comment:
could have been `out_pdf *= pdf`

also good to call it `localPdf` instead of `pdf`